File: python-sdk/arc402/agreement.py
"""ServiceAgreementClient — interacts with the on-chain ServiceAgreement contract."""

from __future__ import annotations

from typing import TYPE_CHECKING

from web3 import Web3

from .abis import ServiceAgreement_ABI
from .types import (
    Agreement,
    DirectDisputeReason,
    DisputeCase,
    DisputeEvidence,
    DisputeOutcome,
    EvidenceType,
    ProviderResponseType,
    RemediationCase,
    RemediationFeedback,
    RemediationResponse,
)

if TYPE_CHECKING:
    from web3.contract import Contract
    from eth_account.signers.local import LocalAccount

ZERO_ADDRESS = "0x0000000000000000000000000000000000000000"


class ServiceAgreementClient:
    """Python wrapper for the ARC-402 ServiceAgreement contract."""

    def __init__(self, address: str, w3: Web3, account: "LocalAccount | None" = None):
        self._w3 = w3
        self._account = account
        self._contract: "Contract" = w3.eth.contract(
            address=Web3.to_checksum_address(address),
            abi=ServiceAgreement_ABI,
        )

    async def propose(
        self,
        provider: str,
        service_type: str,
        description: str,
        price: int,
        token: str,
        deadline: int,
        deliverables_hash: str,
    ) -> tuple[int, str]:
        self._require_account()
        dh_bytes = self._to_bytes32(deliverables_hash)
        is_eth = token == ZERO_ADDRESS or token is None
        tx_params = self._tx_params()
        if is_eth:
            tx_params["value"] = price

        tx = self._contract.functions.propose(
            Web3.to_checksum_address(provider),
            service_type,
            description,
            price,
            Web3.to_checksum_address(token) if token else ZERO_ADDRESS,
            deadline,
            dh_bytes,
        ).build_transaction(tx_params)

        receipt = await self._send(tx)
        return self._extract_agreement_id(receipt), receipt["transactionHash"].hex()

    async def accept(self, agreement_id: int) -> str:
        return await self._simple_write("accept", agreement_id)

    async def fulfill(self, agreement_id: int, actual_deliverables_hash: str) -> str:
        return await self._simple_write("fulfill", agreement_id, self._to_bytes32(actual_deliverables_hash))

    async def commit_deliverable(self, agreement_id: int, deliverable_hash: str) -> str:
        return await self._simple_write("commitDeliverable", agreement_id, self._to_bytes32(deliverable_hash))

    async def verify_deliverable(self, agreement_id: int) -> str:
        return await self._simple_write("verifyDeliverable", agreement_id)

    async def auto_release(self, agreement_id: int) -> str:
        return await self._simple_write("autoRelease", agreement_id)

    async def dispute(self, agreement_id: int, reason: str) -> str:
        return await self._simple_write("dispute", agreement_id, reason)

    async def direct_dispute(self, agreement_id: int, direct_reason: DirectDisputeReason, reason: str) -> str:
        return await self._simple_write("directDispute", agreement_id, int(direct_reason), reason)

    async def request_revision(
        self,
        agreement_id: int,
        feedback_hash: str,
        feedback_uri: str = "",
        previous_transcript_hash: str | bytes | None = None,
    ) -> str:
        return await self._simple_write(
            "requestRevision",
            agreement_id,
            self._to_bytes32(feedback_hash),
            feedback_uri,
            self._to_bytes32(previous_transcript_hash),
        )

    async def respond_to_revision(
        self,
        agreement_id: int,
        response_type: ProviderResponseType,
        proposed_provider_payout: int,
        response_hash: str,
        response_uri: str = "",
        previous_transcript_hash: str | bytes | None = None,
    ) -> str:
        return await self._simple_write(
            "respondToRevision",
            agreement_id,
            int(response_type),
            proposed_provider_payout,
            self._to_bytes32(response_hash),
            response_uri,
            self._to_bytes32(previous_transcript_hash),
        )

    async def propose_partial_settlement(
        self,
        agreement_id: int,
        provider_payout: int,
        response_hash: str,
        response_uri: str = "",
        previous_transcript_hash: str | bytes | None = None,
    ) -> str:
        return await self.respond_to_revision(
            agreement_id=agreement_id,
            response_type=ProviderResponseType.PARTIAL_SETTLEMENT,
            proposed_provider_payout=provider_payout,
            response_hash=response_hash,
            response_uri=response_uri,
            previous_transcript_hash=previous_transcript_hash,
        )

    async def request_human_review(
        self,
        agreement_id: int,
        response_hash: str,
        response_uri: str = "",
        previous_transcript_hash: str | bytes | None = None,
    ) -> str:
        return await self.respond_to_revision(
            agreement_id=agreement_id,
            response_type=ProviderResponseType.REQUEST_HUMAN_REVIEW,
            proposed_provider_payout=0,
            response_hash=response_hash,
            response_uri=response_uri,
            previous_transcript_hash=previous_transcript_hash,
        )

    async def escalate_to_dispute(self, agreement_id: int, reason: str) -> str:
        return await self._simple_write("escalateToDispute", agreement_id, reason)

    def can_direct_dispute(self, agreement_id: int, direct_reason: DirectDisputeReason) -> bool:
        return bool(self._contract.functions.canDirectDispute(agreement_id, int(direct_reason)).call())

    async def submit_dispute_evidence(
        self,
        agreement_id: int,
        evidence_type: EvidenceType,
        evidence_hash: str,
        evidence_uri: str = "",
    ) -> str:
        return await self._simple_write(
            "submitDisputeEvidence",
            agreement_id,
            int(evidence_type),
            self._to_bytes32(evidence_hash),
            evidence_uri,
        )

    async def resolve_dispute_detailed(
        self,
        agreement_id: int,
        outcome: DisputeOutcome,
        provider_award: int,
        client_award: int,
    ) -> str:
        return await self._simple_write(
            "resolveDisputeDetailed",
            agreement_id,
            int(outcome),
            provider_award,
            client_award,
        )

    async def cancel(self, agreement_id: int) -> str:
        return await self._simple_write("cancel", agreement_id)

    async def expired_cancel(self, agreement_id: int) -> str:
        return await self._simple_write("expiredCancel", agreement_id)

    async def expired_dispute_refund(self, agreement_id: int) -> str:
        return await self._simple_write("expiredDisputeRefund", agreement_id)

    def get_agreement(self, agreement_id: int) -> Agreement:
        return Agreement.from_raw(self._contract.functions.getAgreement(agreement_id).call())

    def get_agreements_by_client(self, client: str) -> list[int]:
        return list(self._contract.functions.getAgreementsByClient(Web3.to_checksum_address(client)).call())

    def get_agreements_by_provider(self, provider: str) -> list[int]:
        return list(self._contract.functions.getAgreementsByProvider(Web3.to_checksum_address(provider)).call())

    def get_remediation_case(self, agreement_id: int) -> RemediationCase:
        return RemediationCase.from_raw(self._contract.functions.getRemediationCase(agreement_id).call())

    def get_remediation_feedback(self, agreement_id: int, index: int) -> RemediationFeedback:
        return RemediationFeedback.from_raw(self._contract.functions.getRemediationFeedback(agreement_id, index).call())

    def get_remediation_response(self, agreement_id: int, index: int) -> RemediationResponse:
        return RemediationResponse.from_raw(self._contract.functions.getRemediationResponse(agreement_id, index).call())

    def get_dispute_case(self, agreement_id: int) -> DisputeCase:
        return DisputeCase.from_raw(self._contract.functions.getDisputeCase(agreement_id).call())

    def get_dispute_evidence(self, agreement_id: int, index: int) -> DisputeEvidence:
        return DisputeEvidence.from_raw(self._contract.functions.getDisputeEvidence(agreement_id, index).call())

    def agreement_count(self) -> int:
        return self._contract.functions.agreementCount().call()

    def remediation_history(self, agreement_id: int) -> dict[str, list]:
        case = self.get_remediation_case(agreement_id)
        feedback = [self.get_remediation_feedback(agreement_id, i) for i in range(case.cycle_count)]
        responses = []
        for i in range(case.cycle_count):
            try:
                responses.append(self.get_remediation_response(agreement_id, i))
            except Exception:
                break
        return {"case": case, "feedback": feedback, "responses": responses}

    def dispute_evidence_list(self, agreement_id: int) -> list[DisputeEvidence]:
        dispute = self.get_dispute_case(agreement_id)
        return [self.get_dispute_evidence(agreement_id, i) for i in range(dispute.evidence_count)]

    def _extract_agreement_id(self, receipt: dict) -> int:
        try:
            events = self._contract.events.AgreementProposed().process_receipt(receipt)
            if events:
                return int(events[0]["args"]["id"])
        except Exception:
            pass
        for log in receipt.get("logs", []):
            topics = log.get("topics", [])
            if len(topics) >= 2:
                return int.from_bytes(topics[1], "big")
        raise ValueError("ServiceAgreementClient: could not extract agreement ID from receipt")

    async def _simple_write(self, fn_name: str, *args) -> str:
        self._require_account()
        fn = getattr(self._contract.functions, fn_name)(*args)
        receipt = await self._send(fn.build_transaction(self._tx_params()))
        return receipt["transactionHash"].hex()

    def _require_account(self) -> None:
        if self._account is None:
            raise ValueError("ServiceAgreementClient: an account is required for write methods. Pass account= when constructing the client.")

    def _to_bytes32(self, value: str | bytes | None) -> bytes:
        if value in (None, ""):
            return b"\x00" * 32
        if isinstance(value, bytes):
            return value.ljust(32, b"\x00")[:32]
        return bytes.fromhex(value.removeprefix("0x")).ljust(32, b"\x00")[:32]

    def _tx_params(self) -> dict:
        return {
            "from": self._account.address,
            "nonce": self._w3.eth.get_transaction_count(self._account.address),
            "gas": 500_000,
            "gasPrice": self._w3.eth.gas_price,
            "chainId": self._w3.eth.chain_id,
        }

    async def _send(self, tx: dict) -> dict:
        signed = self._account.sign_transaction(tx)
        tx_hash = self._w3.eth.send_raw_transaction(signed.raw_transaction)
        return self._w3.eth.wait_for_transaction_receipt(tx_hash)

File: reference/sdk/src/agreement.ts
import { ContractRunner, ethers } from "ethers";
import { SERVICE_AGREEMENT_ABI } from "./contracts";
import {
  Agreement,
  AgreementStatus,
  ArbitrationCase,
  ArbitrationVote,
  DisputeCase,
  DisputeClass,
  DisputeEvidence,
  DisputeMode,
  DisputeOutcome,
  DirectDisputeReason,
  EvidenceType,
  ProposeParams,
  ProviderResponseType,
  RemediationCase,
  RemediationFeedback,
  RemediationResponse,
} from "./types";

export class ServiceAgreementClient {
  private contract: ethers.Contract;

  constructor(address: string, runner: ContractRunner) {
    this.contract = new ethers.Contract(address, SERVICE_AGREEMENT_ABI, runner);
  }

  async propose(params: ProposeParams): Promise<{ agreementId: bigint; receipt: ethers.TransactionReceipt }> {
    const isEth = params.token === ethers.ZeroAddress;
    const tx = await this.contract.propose(
      params.provider,
      params.serviceType,
      params.description,
      params.price,
      params.token,
      params.deadline,
      params.deliverablesHash,
      { value: isEth ? params.price : 0n },
    );
    const receipt = await tx.wait();
    const iface = new ethers.Interface(SERVICE_AGREEMENT_ABI);
    for (const log of receipt.logs) {
      try {
        const parsed = iface.parseLog(log);
        if (parsed?.name === "AgreementProposed") return { agreementId: BigInt(parsed.args.id), receipt };
      } catch {}
    }
    throw new Error("Could not parse AgreementProposed event");
  }

  async accept(id: bigint) { const tx = await this.contract.accept(id); return tx.wait(); }
  /** @deprecated Legacy/trusted-only immediate release path. Prefer commitDeliverable() + verifyDeliverable()/autoRelease(). */
  async fulfill(id: bigint, hash: string) { const tx = await this.contract.fulfill(id, hash); return tx.wait(); }
  async fulfillLegacyTrustedOnly(id: bigint, hash: string) { return this.fulfill(id, hash); }
  async commitDeliverable(id: bigint, hash: string) { const tx = await this.contract.commitDeliverable(id, hash); return tx.wait(); }
  async verifyDeliverable(id: bigint) { const tx = await this.contract.verifyDeliverable(id); return tx.wait(); }
  async autoRelease(id: bigint) { const tx = await this.contract.autoRelease(id); return tx.wait(); }
  async dispute(id: bigint, reason: string, feeEth = 0n) { const tx = await this.contract.dispute(id, reason, { value: feeEth }); return tx.wait(); }
  async directDispute(id: bigint, directReason: DirectDisputeReason, reason: string, feeEth = 0n) { const tx = await this.contract.directDispute(id, directReason, reason, { value: feeEth }); return tx.wait(); }
  async escalateToDispute(id: bigint, reason: string, feeEth = 0n) { const tx = await this.contract.escalateToDispute(id, reason, { value: feeEth }); return tx.wait(); }

  /** Open a dispute with explicit mode and class. For ETH agreements, pass fee as feeEth. */
  async openDisputeWithMode(
    id: bigint,
    mode: DisputeMode,
    disputeClass: DisputeClass,
    reason: string,
    feeEth = 0n
  ): Promise<ethers.TransactionReceipt> {
    const tx = await this.contract.openDisputeWithMode(id, mode, disputeClass, reason, { value: feeEth });
    return tx.wait();
  }

  /** Shorthand: UNILATERAL dispute with specified class. */
  async openUnilateralDispute(
    id: bigint,
    disputeClass: DisputeClass,
    reason: string,
    feeEth = 0n
  ): Promise<ethers.TransactionReceipt> {
    return this.openDisputeWithMode(id, DisputeMode.UNILATERAL, disputeClass, reason, feeEth);
  }

  /** Shorthand: MUTUAL dispute — opener pays half. Respondent calls DisputeArbitrationClient.joinMutualDispute(). */
  async openMutualDispute(
    id: bigint,
    disputeClass: DisputeClass,
    reason: string,
    halfFeeEth = 0n
  ): Promise<ethers.TransactionReceipt> {
    return this.openDisputeWithMode(id, DisputeMode.MUTUAL, disputeClass, reason, halfFeeEth);
  }
  async canDirectDispute(id: bigint, directReason: DirectDisputeReason): Promise<boolean> { return this.contract.canDirectDispute(id, directReason); }
  async cancel(id: bigint) { const tx = await this.contract.cancel(id); return tx.wait(); }
  async expiredCancel(id: bigint) { const tx = await this.contract.expiredCancel(id); return tx.wait(); }
  async resolveDisputeDetailed(id: bigint, outcome: DisputeOutcome, providerAward: bigint, clientAward: bigint) {
    const tx = await this.contract.resolveDisputeDetailed(id, outcome, providerAward, clientAward); return tx.wait();
  }

  /** Owner-only: resolve a dispute directly in favor of provider or client. Requires DISPUTED or ESCALATED_TO_HUMAN status. */
  async ownerResolveDispute(agreementId: bigint, favorProvider: boolean): Promise<ethers.TransactionReceipt> {
    const tx = await this.contract.ownerResolveDispute(agreementId, favorProvider);
    return tx.wait();
  }

  /**
   * Called by the DisputeArbitration contract to resolve a dispute with split amounts.
   * @param agreementId Agreement to resolve
   * @param recipient  Address of the winning party
   * @param providerAmount Provider payout in token units
   * @param clientAmount   Client refund in token units
   */
  async resolveFromArbitration(
    agreementId: bigint,
    recipient: string,
    providerAmount: bigint,
    clientAmount: bigint,
  ): Promise<ethers.TransactionReceipt> {
    const tx = await this.contract.resolveFromArbitration(agreementId, recipient, providerAmount, clientAmount);
    return tx.wait();
  }

  async requestRevision(agreementId: bigint, feedbackHash: string, feedbackURI = "", previousTranscriptHash = ethers.ZeroHash) {
    const tx = await this.contract.requestRevision(agreementId, feedbackHash, feedbackURI, previousTranscriptHash);
    return tx.wait();
  }

  async respondToRevision(
    agreementId: bigint,
    responseType: ProviderResponseType,
    responseHash: string,
    responseURI = "",
    previousTranscriptHash = ethers.ZeroHash,
    proposedProviderPayout = 0n,
  ) {
    const tx = await this.contract.respondToRevision(
      agreementId,
      responseType,
      proposedProviderPayout,
      responseHash,
      responseURI,
      previousTranscriptHash,
    );
    return tx.wait();
  }

  async submitDisputeEvidence(agreementId: bigint, evidenceType: EvidenceType, evidenceHash: string, evidenceURI = "") {
    const tx = await this.contract.submitDisputeEvidence(agreementId, evidenceType, evidenceHash, evidenceURI);
    return tx.wait();
  }

  async nominateArbitrator(agreementId: bigint, arbitrator: string) {
    const tx = await this.contract.nominateArbitrator(agreementId, arbitrator);
    return tx.wait();
  }

  async castArbitrationVote(agreementId: bigint, vote: ArbitrationVote, providerAward: bigint, clientAward: bigint) {
    const tx = await this.contract.castArbitrationVote(agreementId, vote, providerAward, clientAward);
    return tx.wait();
  }

  async requestHumanEscalation(agreementId: bigint, reason: string) {
    const tx = await this.contract.requestHumanEscalation(agreementId, reason);
    return tx.wait();
  }

  async getAgreement(id: bigint): Promise<Agreement> {
    const raw = await this.contract.getAgreement(id);
    return {
      id: BigInt(raw.id), client: raw.client, provider: raw.provider, serviceType: raw.serviceType,
      description: raw.description, price: BigInt(raw.price), token: raw.token, deadline: BigInt(raw.deadline),
      deliverablesHash: raw.deliverablesHash, status: Number(raw.status) as AgreementStatus,
      createdAt: BigInt(raw.createdAt), resolvedAt: BigInt(raw.resolvedAt), verifyWindowEnd: BigInt(raw.verifyWindowEnd), committedHash: raw.committedHash,
    };
  }

  async getRemediationCase(id: bigint): Promise<RemediationCase> {
    const raw = await this.contract.getRemediationCase(id);
    return { cycleCount: Number(raw.cycleCount), openedAt: BigInt(raw.openedAt), deadlineAt: BigInt(raw.deadlineAt), lastActionAt: BigInt(raw.lastActionAt), latestTranscriptHash: raw.latestTranscriptHash, active: raw.active };
  }

  async getRemediationFeedback(id: bigint, index: bigint): Promise<RemediationFeedback> {
    const raw = await this.contract.getRemediationFeedback(id, index);
    return { cycle: Number(raw.cycle), author: raw.author, feedbackHash: raw.feedbackHash, feedbackURI: raw.feedbackURI, previousTranscriptHash: raw.previousTranscriptHash, transcriptHash: raw.transcriptHash, timestamp: BigInt(raw.timestamp) };
  }

  async getRemediationResponse(id: bigint, index: bigint): Promise<RemediationResponse> {
    const raw = await this.contract.getRemediationResponse(id, index);
    return { cycle: Number(raw.cycle), author: raw.author, responseType: Number(raw.responseType) as ProviderResponseType, proposedProviderPayout: BigInt(raw.proposedProviderPayout), responseHash: raw.responseHash, responseURI: raw.responseURI, previousTranscriptHash: raw.previousTranscriptHash, transcriptHash: raw.transcriptHash, timestamp: BigInt(raw.timestamp) };
  }

  async getDisputeCase(id: bigint): Promise<DisputeCase> {
    const raw = await this.contract.getDisputeCase(id);
    return { agreementId: BigInt(raw.agreementId), openedAt: BigInt(raw.openedAt), responseDeadlineAt: BigInt(raw.responseDeadlineAt), outcome: Number(raw.outcome) as DisputeOutcome, providerAward: BigInt(raw.providerAward), clientAward: BigInt(raw.clientAward), humanReviewRequested: raw.humanReviewRequested, evidenceCount: BigInt(raw.evidenceCount) };
  }

  async getDisputeEvidence(id: bigint, index: bigint): Promise<DisputeEvidence> {
    const raw = await this.contract.getDisputeEvidence(id, index);
    return { submitter: raw.submitter, evidenceType: Number(raw.evidenceType) as EvidenceType, evidenceHash: raw.evidenceHash, evidenceURI: raw.evidenceURI, timestamp: BigInt(raw.timestamp) };
  }

  async getArbitrationCase(id: bigint): Promise<ArbitrationCase> {
    const raw = await this.contract.getArbitrationCase(id);
    return {
      agreementId: BigInt(raw.agreementId),
      arbitrators: [...raw.arbitrators],
      arbitratorCount: Number(raw.arbitratorCount),
      providerVotes: Number(raw.providerVotes),
      clientVotes: Number(raw.clientVotes),
      splitVotes: Number(raw.splitVotes),
      humanVotes: Number(raw.humanVotes),
      selectionDeadlineAt: BigInt(raw.selectionDeadlineAt),
      decisionDeadlineAt: BigInt(raw.decisionDeadlineAt),
      splitProviderAward: BigInt(raw.splitProviderAward),
      splitClientAward: BigInt(raw.splitClientAward),
      finalized: raw.finalized,
      humanBackstopUsed: raw.humanBackstopUsed,
    };
  }

  async getDisputeEvidenceAll(id: bigint): Promise<DisputeEvidence[]> {
    const dispute = await this.getDisputeCase(id);
    return Promise.all(Array.from({ length: Number(dispute.evidenceCount) }, (_, i) => this.getDisputeEvidence(id, BigInt(i))));
  }

  async getClientAgreements(client: string): Promise<Agreement[]> {
    const ids = await this.contract.getAgreementsByClient(client) as bigint[];
    return Promise.all(ids.map((id) => this.getAgreement(BigInt(id))));
  }

  async getProviderAgreements(provider: string): Promise<Agreement[]> {
    const ids = await this.contract.getAgreementsByProvider(provider) as bigint[];
    return Promise.all(ids.map((id) => this.getAgreement(BigInt(id))));
  }
}
